source/E2E.Tests/Services/BesiegerCamps/BesiegerCampCollectionTests.cs

```diff
@@ -0,0 +1,172 @@
+﻿using E2E.Tests.Environment;
+using E2E.Tests.Environment.Instance;
+using E2E.Tests.Util;
+using GameInterface.Services.BesiegerCamps.Patches;
+using HarmonyLib;
+using System.Reflection;
+using TaleWorlds.CampaignSystem.Party;
+using TaleWorlds.CampaignSystem.Siege;
+using Xunit.Abstractions;
+
+namespace E2E.Tests.Services.BesiegerCamps
+{
+    public class BesiegerCampCollectionTests : IDisposable
+    {
+        private List<MethodBase> disabledMethods = new();
+
+        private E2ETestEnvironment TestEnvironment { get; }
+
+        private EnvironmentInstance Server => TestEnvironment.Server;
+        private IEnumerable<EnvironmentInstance> Clients => TestEnvironment.Clients;
+
+        public BesiegerCampCollectionTests(ITestOutputHelper output)
+        {
+            TestEnvironment = new E2ETestEnvironment(output);
+            DisableMethods();
+        }
+
+        public void Dispose()
+        {
+            TestEnvironment.Dispose();
+        }
+
+        private void DisableMethods()
+        {
+            disabledMethods = new List<MethodBase> {
+                AccessTools.Method(typeof (MobileParty), nameof(MobileParty.OnPartyJoinedSiegeInternal)),
+                AccessTools.Method(typeof (BesiegerCamp), nameof(BesiegerCamp.SetSiegeCampPartyPosition)),
+                AccessTools.Method(typeof (BesiegerCamp), nameof(BesiegerCamp.InitializeSiegeEventSide))
+            };
+        }
+
+        private T ServerCreateObject<T>(out string objectId)
+        {
+            string? id = null;
+            T? obj = default;
+
+            Server.Call(() =>
+            {
+                obj = GameObjectCreator.CreateInitializedObject<T>();
+                Assert.True(Server.ObjectManager.TryGetId(obj, out id));
+            }, disabledMethods);
+
+            objectId = id!;
+            return obj ?? throw new InvalidOperationException("Failed to create object.");
+        }
+
+        [Fact]
+        private void ServerAddBesiegerParty_SyncAllClients()
+        {
+            //Arrange
+            var besiegerCamp = ServerCreateObject<BesiegerCamp>(out string besiegerCampId);
+            var besiegerParty = ServerCreateObject<MobileParty>(out string besiegerPartyId);
+
+            //Act
+            Server.Call(() =>
+            {
+                //besiegerCamp._besiegerParties.Add(besiegerParty); // Doesn't call override, must call it directly
+                BesiegerCampCollectionPatches.ListAddOverride(besiegerCamp._besiegerParties, besiegerParty, besiegerCamp);
```

Comment:
Also check that it is added on the server side

```suggestion
                BesiegerCampCollectionPatches.ListAddOverride(besiegerCamp._besiegerParties, besiegerParty, besiegerCamp);
                Assert.Contains<MobileParty>(besiegerCamp._besiegerParties, besiegerParty);
```

Reply:
Done